edgeless_node/src/resources/ollama.rs

```diff
@@ -0,0 +1,344 @@
+                    _ => {
+                        continue;
+                    }
+                };
+
+                let (reply_sender, reply_receiver) =
+                    tokio::sync::oneshot::channel::<anyhow::Result<(edgeless_api::function_instance::InstanceId, String)>>();
+                let _ = sender
+                    .send(OllamaCommand::Chat(
+                        model_name.clone(),
+                        history_id.clone(),
+                        message_data,
+                        instance_id.function_id.clone(),
+                        reply_sender,
+                    ))
+                    .await;
+
+                match reply_receiver.await {
+                    Ok(response) => match response {
+                        Ok((target, response)) => {
+                            let _ = dataplane_handle.send(target, response).await;
+                        }
+                        Err(err) => {
+                            log::warn!("Error from ollama: {}", err)
+                        }
+                    },
+                    Err(err) => {
+                        log::warn!("Communication error with ollama resource provider: {}", err)
+                    }
+                };
+
+                if need_reply {
+                    dataplane_handle
+                        .reply(source_id, channel_id, edgeless_dataplane::core::CallRet::Reply("".to_string()))
+                        .await;
+                }
+            }
+        });
+
+        Ok(Self { join_handle: handle })
+    }
+}
+
+impl OllamaResourceProvider {
+    /// Create an Ollama resource provider:
+    ///
+    /// - `dataplane_provider`: handle to the EDGELESS data plane
+    /// - `resource_provider_id`: identifier of this resource provider,
+    ///    also containing the identifier of the node hosting it
+    /// - `ollama_host`: address of the ollama server
+    /// - `ollama_port`: port number of the ollama server
+    /// - `ollama_messages_number_limit`: maximum number of messages per
+    ///    chat conversation
+    pub async fn new(
+        dataplane_provider: edgeless_dataplane::handle::DataplaneProvider,
+        resource_provider_id: edgeless_api::function_instance::InstanceId,
+        ollama_host: &str,
+        ollama_port: u16,
+        ollama_messages_number_limit: u16,
+    ) -> Self {
+        // Create a channel for:
+        // - single receiver: the loop in the task below
+        // - multiple senders: the resource instances that will be created
+        //   at run-time
+        let (sender, receiver) = futures::channel::mpsc::unbounded();
+        let mut receiver: futures::channel::mpsc::UnboundedReceiver<OllamaCommand> = receiver;
+
+        // Create a new instance of the ollama connector.
+        let mut ollama = ollama_rs::Ollama::new_with_history(format!("http://{}", ollama_host), ollama_port, ollama_messages_number_limit);
```

Comment:
Can you only have one client?
Otherwise you might just use that from the resource-instance task.

The HTTP Server is different as it needs to multiplex across resource instances.

Reply:
You can have multiple clients. But I _wanted_ to have a single client, by design, so that we can control the scheduling of requests sent to the ollama server. At the moment we have a single task serving in a FIFO manner, but in the future you can change that to multiple tasks or to more complex scheduling policies (e.g., prioritise some kinds of requests, or group requests for the same model together).